### PR TITLE
Switch MSC Code Utils to download csv file lazy when feature is enabled

### PR DESCRIPTION
**Pull request by Siedlerchr**
### Related issues and pull requests

Closes https://github.com/JabRef/jabref-issue-melting-pot/issues/1312

Preferences:
grafik

grafik

grafik

   

### PR Description

* Migrate from static JSON loading to CSV loading at runtime and only when activated in the preferences, as this feature is only relevant for math people. We cannot ship the CSV as it is licensed under CC-BY-SA-NC 
* Wrap the csv file in an mvstore.  
* Before Downloading the user has to acknowledge the license

´´

   

### Steps to test

1. Enable Feature in Prefs
2.  Go to entry editor -> Keywords 
3. add a new Keyword 34B60
4. hover with mouse and see text Applications of boundary value problems involving ordinary differential equations 
   

### AI usage

chat gpt codex 5.4 mini and gemini flash 3 
Reviewd the code myself 

_____

   

### Checklist

   

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] If AI tools were used, I disclosed them in the "AI usage" section and reviewed, understood, and take full ownership of all AI-generated code
- [x] I manually tested my changes in running JabRef (always required)
- [x] I added JUnit tests for changes (if applicable)
- [x] I added screenshots in the PR description (if change is visible to the user)
- [/] I added a screenshot in the PR description showing a library with a single entry with me as author and as title the issue number
- [x] I described the change in `CHANGELOG.md` in a way that can be understood by the average user (if change is visible to the user)
- [.] I checked the [user documentation](https://docs.jabref.org/) for up to dateness and submitted a pull request to our [user documentation repository](https://github.com/JabRef/user-documentation/tree/main/en)

### CODE REVIEW

File: jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java [hunk 1]
```diff
@@ -104,11 +75,12 @@ private boolean isSeparatedKeyword(String keywordString, String keywordSeparator
 
     @Override
     protected void customizeTag(Label tagLabel, Keyword keyword) {
-        if (mscmap.containsKey(tagLabel.getText())) {
-            String mscClassification = mscmap.get(tagLabel.getText());
-            Tooltip tooltip = new Tooltip(mscClassification);
-            tagLabel.setTooltip(tooltip);
+        if (!preferences.shouldEnableMscKeywordDescriptions()) {
```
Comment (koppor) [inline jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java]: Can we just store these preferences locally? (I think we also do that elsewhere)

File: jabgui/src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java [hunk 2]
```diff
@@ -556,6 +580,26 @@ public void openBrowser() {
...
+    private void downloadMscCodes() {
+        Path mscMvFile = Directories.getMscDirectory().resolve(MscCodeLoader.MSC_FILE_NAME);
+        if (MscCodeLoader.isMvStoreAvailableWithData(mscMvFile)) {
+            return;
+        }
+
+        dialogService.notify(Localization.lang("Downloading MSC codes..."));
+
+        BackgroundTask.wrap(() -> {
+                          MscCodeLoader.downloadAndConvert(URLUtil.create(MscCodeLoader.MSC_CSV_URL), mscMvFile);
```
Comment (koppor) [inline jabgui/src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java]: Too much load. 😅

File: jablib/src/test/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatterTest.java [hunk 3]
```diff
@@ -33,14 +55,14 @@ void convertSingleMSCCode() {
...
-                "53C99 None of the above but in this section,Biopropulsion in water and in air,Search theory",
+                "None of the above\\, but in this section,Biopropulsion in water and in air,Search theory",
                 formatter.format("53C99,76Z10,90B40"));
     }
 
     @Test
     void preserveNonMSCKeyword() {
         assertEquals(
-                "Jabref123,Search theory,Hello Jabref,Direct methods ( G-spaces of Busemann etc.)",
+                "Jabref123,Search theory,Hello Jabref,Direct methods (\\\\(G\\\\)-spaces of Busemann\\, etc.)",
```
Comment (koppor) [inline jablib/src/test/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatterTest.java:69]: The test is about NON-MSC keywords preserved - why does JabRef change them?

**Review by koppor — DISMISSED** [review]
small changes

The file ´CC-BY-SA-NC` cannot be downloaded too just automically.

The UI needs to have a warning and clearly link the license. We also need to cache the file.

A new dir needs to be added at org.jabref.logic.util.Directories - maybe "mvcodes"? There, the csv and the .mv can live in.

File: jablib/src/test/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatterTest.java (hunk 3 above)
Comment (Siedlerchr) [reply jablib/src/test/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatterTest.java:69]: the non MSC keyword is JabRef 123. The others are valid msc codes and expected to be formatted. I renamed the test to make it clear. 
```
"53C70"	"Direct methods (\(G\)-spaces of Busemann, etc.)"	"Direct methods (\(G\)-spaces of Busemann, etc.)"
```

File: jablib/src/main/java/org/jabref/logic/util/Directories.java [hunk 4]
```diff
@@ -75,4 +75,13 @@ public static Path getCoverDirectory() {
                                       "covers",
                                       OS.APP_DIR_APP_AUTHOR));
     }
+
+    public static Path getMscDirectory() {
+        return Path.of(
+                AppDirsFactory.getInstance()
+                              .getUserDataDir(
```
Comment (Siedlerchr) [inline jablib/src/main/java/org/jabref/logic/util/Directories.java:82]: @koppor  I have a dir for the file

File: jabgui/src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java (hunk 2 above)
Comment (Siedlerchr) [reply jabgui/src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java]: ??

File: jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java (hunk 1 above)
Comment (Siedlerchr) [reply jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java]: Talked to @calixtus  and moved to Abbreviaton preferences

File: jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java
Comment (subhramit) [inline jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java]: For future - `KeywordsEditorViewModel` uses only `preferences.getBibEntryPreferences().getKeywordSeparator()` so we can improve abstraction instead of passing entire clipreferences.

File: jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java
Comment (subhramit) [inline jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java]: IntelliJ says Static member 'org.jabref.gui.entryeditor.EntryEditorPreferences.getDefaultEntryEditorTabs()' accessed via instance reference

File: jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java [hunk 5]
```diff
@@ -73,10 +92,12 @@ public void setValues() {
         journalPopupProperty.setValue(entryEditorPreferences.shouldEnableJournalPopup() == EntryEditorPreferences.JournalPopupEnabled.ENABLED);
         autoLinkEnabledProperty.setValue(entryEditorPreferences.autoLinkFilesEnabled());
         enableSciteTabProperty.setValue(entryEditorPreferences.shouldShowSciteTab());
+        enableMscKeywordDescriptionsProperty.setValue(abbreviationPreferences.shouldEnableMscKeywordDescriptions());
         showUserCommentsProperty.setValue(entryEditorPreferences.shouldShowUserCommentsFields());
         citationCountFetcherTypeProperty.setValue(entryEditorPreferences.getCitationCountFetcherType());
 
         setFields(entryEditorPreferences.getEntryEditorTabs());
+        mscKeywordDescriptionsInitialized = true;
```
Comment (subhramit) [inline jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java:101]: Do we need this?

File: jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java [hunk 6]
```diff
@@ -205,11 +227,59 @@ public BooleanProperty enableSciteTabProperty() {
...
+            if (accepted) {
+                downloadMscCodes();
+            } else {
+                enableMscKeywordDescriptionsProperty.setValue(false);
+            }
+        }
+    }
+
+    private void downloadMscCodes() {
+        var mscMvFile = Directories.getMscDirectory().resolve(MscCodeLoader.MSC_FILE_NAME);
```
Comment (subhramit) [inline jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java]: nit
```suggestion
        Path mscMvFile = Directories.getMscDirectory().resolve(MscCodeLoader.MSC_FILE_NAME);
```
and add import

File: jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java [hunk 7]
```diff
@@ -1,89 +1,57 @@
...
+import org.jspecify.annotations.Nullable;
 
 public class ConvertMSCCodesFormatter extends Formatter implements LayoutFormatter {
-    private static final Logger LOGGER = LoggerFactory.getLogger(ConvertMSCCodesFormatter.class);
-    private static final BiMap<String, String> MSCMAP;
-    private static boolean conversionPossible;
 
-    static {
-        MSCMAP = initializeMap();
+    private final @Nullable JabRefCliPreferences cliPreferences;
```
Comment (subhramit) [inline jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:22]: If only `BibEntryPreferences` are used, can we avoid entire clipreferences here as well?

File: jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java [hunk 8]
```diff
@@ -107,4 +75,16 @@ public String getKey() {
...
         return "06E30";
     }
+
+    private JabRefCliPreferences getCliPreferences() {
+        if (cliPreferences != null) {
+            return cliPreferences;
+        }
+        return Injector.instantiateModelOrService(JabRefCliPreferences.class);
+    }
+
```
Comment (subhramit) [inline jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:85]: If used only within this class, I think the constructor injection and data member is enough? Or is this method needed?

**Review by subhramit — COMMENTED** [review]
Some questions/comments

File: jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java (hunk 8 above)
Comment (Siedlerchr) [reply jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:85]: Yes, we cannot pass down the Preferences via constructor, that is currently not possible due to the architecutre of our formatter and layout stuff

File: jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java (hunk 5 above)
Comment (Siedlerchr) [reply jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java:101]: This is to prevent the side effect because Easybind.subscribe also fires on the creation when it's initialized and thus we would always get false

File: jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java
Comment (Siedlerchr) [reply jabgui/src/main/java/org/jabref/gui/preferences/entryeditor/EntryEditorTabViewModel.java]: Yeah, I changed this as well

File: jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java
Comment (Siedlerchr) [reply jabgui/src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java]: good point, passed now the bib entry prefs

**Review by subhramit — DISMISSED** [review]

File: jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java (hunk 7 above)
Comment (Siedlerchr) [reply jablib/src/main/java/org/jabref/logic/formatter/bibtexfields/ConvertMSCCodesFormatter.java:22]: Yeah, we cannot instantiate  sub preferenes via the injector

**Review by subhramit — DISMISSED** [review]

**Review by subhramit — APPROVED** [review]
